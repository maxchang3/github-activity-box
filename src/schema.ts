import { z } from 'zod'

const BasicSearchNodeSchema = z.object({
    type: z.enum(['Issue', 'PullRequest']),
    title: z.string(),
    repo: z.object({
        name: z.string(),
        owner: z.object({
            login: z.string(),
        }),
    }),
    number: z.number(),
    createdAt: z.coerce.date(),
    closedAt: z.nullable(z.coerce.date()),
})

const IssueNodeSchema = BasicSearchNodeSchema.extend({
    type: z.literal('Issue'),
    issueState: z.enum(['OPEN', 'CLOSED']),
})

const PullRequestNodeSchema = BasicSearchNodeSchema.extend({
    type: z.literal('PullRequest'),
    prState: z.enum(['OPEN', 'CLOSED', 'MERGED']),
})

export type IssueNode = z.infer<typeof IssueNodeSchema>
export type PullRequestNode = z.infer<typeof PullRequestNodeSchema>

// To keep the type non-expanded
export type SearchNode = IssueNode | PullRequestNode

const SearchNodeSchema: z.ZodType<SearchNode> = z.discriminatedUnion('type', [
    IssueNodeSchema,
    PullRequestNodeSchema,
])

export const SearchResponseSchema = z.object({
    search: z.object({
        edges: z.array(
            z.object({
                node: SearchNodeSchema,
            })
        ),
    }),
})

export type SearchResponse = z.infer<typeof SearchResponseSchema>
