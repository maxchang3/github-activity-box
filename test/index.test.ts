import { SearchResponseSchema } from '@/schema'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { Issues, PRs } from '~/fixtures'
const mockedUpdate = vi.fn()
const mockedConsoleError = vi.fn()

vi.spyOn(console, 'log').mockImplementation(() => {})
vi.spyOn(console, 'error').mockImplementation(mockedConsoleError)
// @ts-expect-error no need to throw
vi.spyOn(process, 'exit').mockImplementation(() => {})

vi.mock('gist-box', async (importOriginal) => {
    const actual = await importOriginal<typeof import('gist-box')>()
    return {
        ...actual,
        GistBox: vi.fn().mockImplementation(() => ({
            update: mockedUpdate,
        })),
    }
})

vi.mock('@/graphql', async () => ({
    getIssues: vi.fn().mockResolvedValue(
        SearchResponseSchema.parse({
            search: {
                edges: Issues,
            },
        })
    ),
    getPRs: vi.fn().mockResolvedValue(
        SearchResponseSchema.parse({
            search: {
                edges: PRs,
            },
        })
    ),
}))

describe('activity-box', () => {
    beforeEach(async () => {
        mockedUpdate.mockClear()
        vi.stubEnv('EXCLUDE_REPO', '')
        vi.stubEnv('EXCLUDE_OWNER', '')
        vi.stubEnv('ACTIVITY_TYPE', 'all')
    })

    const runAction = async () => {
        vi.resetModules()
        return await import('../src/index')
    }

    it('should update the Gist with the expected string', async () => {
        await runAction()
        expect(mockedUpdate).toHaveBeenCalledOnce()
        expect(mockedUpdate.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should exclude repos listed in `EXCLUDE_REPO`', async () => {
        vi.stubEnv('EXCLUDE_REPO', 'clippy/should-be-filtered')

        await runAction()
        expect(mockedUpdate).toHaveBeenCalledOnce()
        expect(mockedUpdate.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should exclude owners listed in `EXCLUDE_OWNER`', async () => {
        vi.stubEnv('EXCLUDE_OWNER', 'clippy')

        await runAction()
        expect(mockedUpdate).toHaveBeenCalledOnce()
        expect(mockedUpdate.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should respect the `ACTIVITY_TYPE` filter', async () => {
        vi.stubEnv('ACTIVITY_TYPE', 'issue')
        await runAction()
        expect(mockedUpdate).toHaveBeenCalledOnce()
        expect(mockedUpdate.mock.calls[0][0]).toMatchSnapshot()
        expect(mockedUpdate.mock.calls[0][0]).not.toContain('PR')
    })

    it('should handle failure to update the Gist', async () => {
        mockedUpdate.mockImplementationOnce(() => {
            throw new Error('404')
        })

        await runAction()
        expect(mockedConsoleError).toHaveBeenCalledOnce()
        expect(mockedConsoleError.mock.calls[0][0]).toMatchSnapshot()
        expect(mockedUpdate).toHaveBeenCalled()
    })
})
