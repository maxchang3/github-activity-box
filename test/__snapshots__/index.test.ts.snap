// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`activity-box > should exclude owners listed in \`EXCLUDE_OWNER\` 1`] = `
{
  "content": "",
  "description": undefined,
}
`;

exports[`activity-box > should exclude repos listed in \`EXCLUDE_REPO\` 1`] = `
{
  "content": "🎉 Merged: PR #4 in clippy/take-over-github
🎉 Merged: PR #32 in clippy/take-over-github
❗️ Open: issue #35 in clippy/really-really-really-really-rea...
❗️ Closed: issue #27 in clippy/take-over-github
💪 Open: PR #38 in clippy/take-over-github",
  "description": undefined,
}
`;

exports[`activity-box > should handle failure to update the Gist 1`] = `"Error getting or update the Gist: Error: 404"`;

exports[`activity-box > should update the Gist with the expected string 1`] = `
{
  "content": "🎉 Merged: PR #4 in clippy/take-over-github
🎉 Merged: PR #32 in clippy/take-over-github
❗️ Open: issue #35 in clippy/really-really-really-really-rea...
❗️ Closed: issue #27 in clippy/take-over-github
💪 Open: PR #50 in clippy/should-be-filtered",
  "description": undefined,
}
`;
